exec/java-exec/src/main/java/org/apache/drill/exec/store/avro/AvroBatchReader.java

```diff
@@ -0,0 +1,368 @@
+    }
+    return "AvroBatchReader[File=" + filePath
+      + ", Position=" + currentPosition
+      + "]";
+  }
+
+  /**
+   * Initialized Avro data reader based on given file system and file path.
+   * Moves reader to the sync point from where to start reading the data.
+   *
+   * @param fileSplit file split
+   * @param fs file system
+   * @param opUserName name of the user whom to impersonate while reading the data
+   * @param queryUserName name of the user who issues the query
+   * @return Avro file reader
+   */
+  private DataFileReader<GenericContainer> prepareReader(FileSplit fileSplit, FileSystem fs, String opUserName, String queryUserName) {
+    try {
+      UserGroupInformation ugi = ImpersonationUtil.createProxyUgi(opUserName, queryUserName);
+      DataFileReader<GenericContainer> reader = ugi.doAs((PrivilegedExceptionAction<DataFileReader<GenericContainer>>) () ->
+        new DataFileReader<>(new FsInput(fileSplit.getPath(), fs.getConf()), new GenericDatumReader<GenericContainer>()));
+
+      // move to sync point from where to read the file
+      reader.sync(fileSplit.getStart());
+      return reader;
+    } catch (IOException | InterruptedException e) {
+      throw UserException.dataReadError(e)
+        .message("Error preparing Avro reader")
+        .addContext("Reader", this)
+        .build(logger);
+    }
+  }
+
+  private boolean nextLine(RowSetLoader rowWriter) {
+    try {
+      if (!reader.hasNext() || reader.pastSync(endPosition)) {
+        return false;
+      }
+      container = reader.next(container);
+    } catch (IOException e) {
+      throw UserException.dataReadError(e)
+        .addContext("Reader", this)
+        .build(logger);
+    }
+
+    Schema schema = container.getSchema();
+    GenericRecord record = (GenericRecord) container;
+
+    if (Schema.Type.RECORD != schema.getType()) {
+      throw UserException.dataReadError()
+        .message("Root object must be record type. Found: %s", schema.getType())
+        .addContext("Reader", this)
+        .build(logger);
+    }
+
+    rowWriter.start();
+    List<Schema.Field> fields = schema.getFields();
+    for (Schema.Field field : fields) {
+      String fieldName = field.name();
+      Object value = record.get(fieldName);
+      ObjectWriter writer = rowWriter.column(fieldName);
+      processRecord(writer, value, field.schema());
+    }
+    rowWriter.save();
+    return true;
+  }
+
+  private void processRecord(ObjectWriter writer, Object value, Schema schema) {
+    // skip processing record if it is null or is not projected
+    if (value == null || !writer.isProjected()) {
+      return;
+    }
+
+    switch (schema.getType()) {
+      case UNION:
+        processRecord(writer, value, AvroSchemaUtil.extractSchemaFromNullable(schema, writer.schema().name()));
+        break;
+      case RECORD:
+        TupleWriter tupleWriter = writer.tuple();
+
+        if (tupleWriter.tupleSchema().isEmpty()) {
```

Comment:
Is this known up front? Can we build up the nested structure when we build the Drill-side schema?

Reply:
For most of the cases yes and we build, we cannot build it though for the case when recursive named record types are present.

Reply:
Do you mean nested record types? Wouldn't those translate to Drill STRUCT (MAP)s? Looking at [this spec](https://avro.apache.org/docs/current/spec.html#schemas), it would seen that an Avro record maps to a Drill tuple (that is, a top-level record or a MAP). Or am I missing something?

Reply:
No, I mean records that refer to named records that are not constructed yet.
Regular, maps / nested map are constructed easily, since we know the structure beforehand:
https://github.com/arina-ielchiieva/drill/blob/DRILL-7454/exec/java-exec/src/test/java/org/apache/drill/exec/store/avro/AvroSchemaUtilTest.java#L246

The case I am referring to is a little bit more complex, in Avro documentation they just provide one example of it:
```
For example, a linked-list of 64-bit values may be defined with:
{
  "type": "record",
  "name": "LongList",
  "aliases": ["LinkedLongs"],                      // old name for this
  "fields" : [
    {"name": "value", "type": "long"},             // each element has a long
    {"name": "next", "type": ["null", "LongList"]} // optional next element
  ]
}
```
Simply put they allow a field of record type to refer to another record by name. The problem occurs when one record type refers to another which is not yet constructed, moreover it cannot be constructed until upper type is constructed.
Some more complicated schema examples: https://github.com/arina-ielchiieva/drill/blob/DRILL-7454/exec/java-exec/src/test/java/org/apache/drill/exec/store/avro/AvroSchemaUtilTest.java#L343

For these cases, I put off such type construction till the runtime when exact schema is know, simply adding map without fields.
Explanation in the code: https://github.com/arina-ielchiieva/drill/blob/DRILL-7454/exec/java-exec/src/main/java/org/apache/drill/exec/store/avro/AvroSchemaUtil.java#L118

In reality, I believe it is a rare case but we had unit test in Drill so I had to cover this case as well.